src/openloco/ui/WindowManager.cpp

```diff
@@ -541,36 +586,292 @@ namespace openloco::ui::WindowManager
+            }
+        } while (loop);
+
+        return createWindowOnScreen(type, position, size, flags, events);
     }
 
-    // 0x004C9F5D
+    /**
+     * 0x004C9F5D
+     *
+     * @param type @<cl>
+     * @param origin @<eax>
+     * @param size @<ebx>
+     * @param flags @<ecx << 8>
+     * @param events @<edx>
+     * @return
+     */
     window* createWindow(
         WindowType type,
         gfx::point_t origin,
         gfx::ui_size_t size,
         uint32_t flags,
         window_event_list* events)
     {
-        registers regs;
-        regs.eax = (origin.y << 16) | (origin.x & 0xFFFF);
-        regs.ebx = (size.height << 16) | (size.width & 0xFFFF);
-        regs.ecx = (uint8_t)type | (flags << 8);
-        regs.edx = (int32_t)events;
-        call(0x004C9F5D, regs);
-        return (window*)regs.esi;
+        if (count() == max_windows)
+        {
+            for (ui::window* w = &_windows[0]; w != _windowsEnd; w++)
+            {
+                if ((w->flags & window_flags::stick_to_back) != 0)
+                    continue;
+
+                if ((w->flags & window_flags::stick_to_front) != 0)
+                    continue;
+
+                if ((w->flags & window_flags::no_auto_close) != 0)
+                    continue;
+
+                close(w);
+                return createWindow(type, origin, size, flags, events);
+            }
+        }
+
+        bool stickToBack = (flags & window_flags::stick_to_back) != 0;
+        bool stickToFront = (flags & window_flags::stick_to_front) != 0;
+        bool hasFlag12 = (flags & window_flags::flag_12) != 0;
+        bool hasFlag13 = (flags & window_flags::flag_13) != 0;
+
+        // Find right position to insert new window
+        size_t dstIndex = 0;
+        if (stickToBack)
+        {
+            for (size_t i = 0; i < count(); i++)
+            {
+                if ((_windows[i].flags & window_flags::stick_to_back) != 0)
+                {
+                    dstIndex = i;
+                }
+            }
+        }
+        else if (stickToFront)
+        {
+            dstIndex = count();
+        }
+        else
+        {
+            for (int i = (int)count(); i > 0; i--)
+            {
+                if ((_windows[i - 1].flags & window_flags::stick_to_front) == 0)
+                {
+                    dstIndex = i;
+                    break;
+                }
+            }
+        }
+
+        auto window = ui::window(origin, size);
+        window.type = type;
+        window.flags = flags;
+        if (hasFlag12
+            || (!stickToBack && !stickToFront && !hasFlag13))
+        {
+            window.flags |= window_flags::white_border_mask;
+            audio::play_sound(audio::sound_id::open_window, origin.x + size.width / 2);
+        }
+
+        window.event_handlers = events;
+
+        size_t length = _windowsEnd - (_windows + dstIndex);
+        memmove(_windows + dstIndex + 1, _windows + dstIndex, length * sizeof(ui::window));
```

Comment:
length can be 0, is this an issue?

Reply:
No, it should just be a no-op if the size argument is 0.